### ISSUE TITLE
How to retrieve basic block predecessors/successors?

### ISSUE THREAD

**Issue by ZornsLemma** [revng/llvmcpy#5]
Hi,

This project seems like a really cool idea, and exactly what I'm looking for - I'd like to experiment with writing a code generator and if I can do it in Python that will help me learn the LLVM API, and I can concentrate on my code generation algorithms without getting bogged down in C++ issues.

However, I can't see how to determine the predecessors/successors of a basic block. Is this possible? Here's what I've tried:

```$ python
Python 2.7.12 (default, Nov 19 2016, 06:48:10) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from llvmcpy.llvm import *
>>> buffer = create_memory_buffer_with_contents_of_file("foo.bc")
>>> context = get_global_context()
>>> module = context.parse_ir(buffer)
>>> function = module.iter_functions().next()
>>> bb = function.iter_basic_blocks().next()
>>> dir(bb)
['__class__', '__delattr__', '__dict__', '__doc__', '__format__', '__getattribute__', '__hash__', '__init__', '__module__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', 'as_value', 'delete', 'first_instruction', 'get_as_value', 'get_first_instruction', 'get_last_instruction', 'get_next', 'get_parent', 'get_previous', 'get_terminator', 'in_ptr', 'insert', 'iter_instructions', 'last_instruction', 'move_basic_block_after', 'move_basic_block_before', 'next', 'out_ptr', 'parent', 'previous', 'ptr', 'remove_basic_block_from_parent', 'terminator']
# no obvious predecessor/successor members of bb
>>> v=bb.get_as_value()
>>> v.get_successor(0)
Segmentation fault (core dumped)
```

Or:
```
>>> bb.get_next().first_instruction().dump()
Segmentation fault (core dumped)
```

Any advice would be appreciated! I've attached foo.bc as a zip file for reference.
[foo.bc.zip](https://github.com/revng/llvmcpy/files/736289/foo.bc.zip)

Cheers.

Steve

**Comment by aleclearmind** [comment]
Have you read the README? You can use the `iter_instructions` method to iterate over the contents of a basic block. `bb.get_as_value` and then `v.get_successor` don't make much sense. If you're not sure what a method does, you can use `help (v.get_successor)` and you should find the name of the underlying LLVM C API being used, then you can lookup what that function does in the LLVM documentation.

Anyway, this is also a bug on our side, we shouldn't allow NULL pointers to be dereferenced. I'll come up with a fix for this.

**Comment by ZornsLemma** [comment]
Thanks. Doesn't iter_instructions iterate over the instructions *within* a basic block? Given a basic block, I would like to know which other basic block(s) are its predecessors and its successors.

**Comment by Oloshka** [comment]
You can get successors of the terminator instruction.
```
succ_bb = bb.get_terminator().get_successor(0)
```

**Comment by ZornsLemma** [comment]
That's great, thanks!

**Comment by aleclearmind** [comment]
> Anyway, this is also a bug on our side, we shouldn't allow NULL pointers to be dereferenced. I'll come up with a fix for this.

Should be fixed in a6b8332. Closing this issue.
